File: lim/이진탐색/Parametric Search - 높이 차를 최소화 하기.py
'''
높이차를 0~500에 대해 이진탐색
검증함수는 최소높이를 1~(500-높이차) 까지 반복하면서 최소최대 높이가 정해진 상태에서 bfs
검증함수는 500 * 100 * 100 시간 걸린다
총 log(500) * 5,000,000

이진탐색 특유의 극단적으로 큰범위가 없고(ex - 10**9)
500이라는 작은 수를 단축하는게 키포인트라서 이진탐색임을 파악하기 힘들다
'''

import sys
from collections import deque


def sol(n, m, grid):
    def in_range(r, c):
        return r >= 0 and c >= 0 and r < n and c < m

    def bfs(min_num, max_num):
        if grid[0][0] < min_num or grid[0][0] > max_num:
            return False

        drs = [1, -1, 0, 0]
        dcs = [0, 0, 1, -1]
        dq = deque([(0, 0)])
        visited = set([(0, 0)])
        while len(dq):
            r, c = dq.popleft()
            if (r, c) == (n-1, m-1):
                return True
            for dr, dc in zip(drs, dcs):
                nr, nc = r + dr, c + dc
                if not in_range(nr, nc) or (nr, nc) in visited:
                    continue
                if grid[nr][nc] < min_num or grid[nr][nc] > max_num:
                    continue
                visited.add((nr, nc))
                dq.append((nr, nc))
        return False

    def possible(diff):
        for min_num in range(1, 501 - diff):
            if bfs(min_num, min_num + diff):
                return True
        return False

    def search():
        left, right = 0, 500
        answer = -1
        while left <= right:
            mid = (left + right) // 2
            if possible(mid):
                right = mid - 1
                answer = mid
            else:
                left = mid + 1
        return answer

    print(search())


n, m = map(int, sys.stdin.readline().split())
grid = [list(map(int, sys.stdin.readline().split())) for _ in range(n)]

sol(n, m, grid)

File: lim/이진탐색/Parametric_Search_높이_차를_최소화_하기.java
package lim.이진탐색;

import java.util.*;
import java.io.*;

public class Parametric_Search_높이_차를_최소화_하기 {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static int n, m;
    static int[] drs = { 1, 0, -1, 0 }, dcs = { 0, 1, 0, -1 };
    static int[][] grid;
    static boolean[][] visited;

    static class Tuple {
        int r, c;

        Tuple(int r, int c) {
            this.r = r;
            this.c = c;
        }
    }

    static boolean inRange(int r, int c) {
        return r >= 0 && c >= 0 && r < n && c < m;
    }

    static boolean canGo(int nr, int nc, int minVal, int maxVal) {
        if (!inRange(nr, nc)) {
            return false;
        }
        return minVal <= grid[nr][nc] && grid[nr][nc] <= maxVal;
    }

    static boolean bfs(int r, int c, int minVal, int maxVal) {
        LinkedList<Tuple> queue = new LinkedList<>();
        visited = new boolean[n][m];
        queue.offer(new Tuple(r, c));
        visited[r][c] = true;

        while (!queue.isEmpty()) {
            Tuple tuple = queue.poll();
            if (tuple.r == n - 1 && tuple.c == m - 1) {
                return true;
            }

            for (int i = 0; i < dcs.length; i++) {
                int nr = tuple.r + drs[i];
                int nc = tuple.c + dcs[i];
                if (!canGo(nr, nc, minVal, maxVal) || visited[nr][nc]) {
                    continue;
                }
                queue.add(new Tuple(nr, nc));
                visited[nr][nc] = true;
            }
        }

        return false;
    }

    static boolean check(int maxDiff) {
        for (int minVal = 0; minVal + maxDiff <= 500; minVal++) {
            int maxVal = minVal + maxDiff;
            if (!canGo(0, 0, minVal, maxVal)) {
                continue;
            }
            if (bfs(0, 0, minVal, maxVal)) {
                return true;
            }
        }
        return false;
    }

    static void sol() throws IOException {
        int left = 0, right = 500;
        int answer = 0;
        while (left <= right) {
            int mid = (left + right) / 2;
            if (check(mid)) {
                right = mid - 1;
                answer = mid;
            } else {
                left = mid + 1;
            }
        }
        bw.write(answer + "");
    }

    public static void main(String[] args) throws IOException {
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        grid = new int[n][m];
        for (int i = 0; i < grid.length; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < grid[0].length; j++) {
                grid[i][j] = Integer.parseInt(st.nextToken());
            }
        }
        sol();
        br.close();
        bw.flush();
        bw.close();
    }
}
